Examples/Framework/src/EventData/NeuralCalibrator.cpp

```diff
@@ -0,0 +1,141 @@
+  int diff0 = icntr0 - 3;
+  int diff1 = icntr1 - 3;
+
+  // Fill the matrix
+  std::array<float, 7 * 7> mat{};
+  for (const ActsExamples::Cluster::Cell& cell : cl.channels) {
+    int im = cell.bin[0] - diff0;
+    int jm = cell.bin[1] - diff1;
+    if (im >= 0 && im < 7 && jm >= 0 && jm < 7)
+      mat.at(im * 7 + jm) = cell.activation;
+  }
+
+  return mat;
+}
+
+}  // namespace detail
+
+ActsExamples::NeuralCalibrator::NeuralCalibrator(
+    const std::filesystem::path& modelPath, size_t nComp,
+    std::vector<size_t> volumeIds)
+    : m_env(ORT_LOGGING_LEVEL_WARNING, "NeuralCalibrator"),
+      m_model(m_env, modelPath.c_str()),
+      m_nComp{nComp},
+      m_volumeIds{std::move(volumeIds)} {}
+
+void ActsExamples::NeuralCalibrator::calibrate(
+    const MeasurementContainer& measurements, const ClusterContainer* clusters,
+    const Acts::GeometryContext& gctx,
+    Acts::MultiTrajectory<Acts::VectorMultiTrajectory>::TrackStateProxy&
+        trackState) const {
+  Acts::SourceLink usl = trackState.getUncalibratedSourceLink();
+  const IndexSourceLink& sourceLink = usl.get<IndexSourceLink>();
+  assert((sourceLink.index() < measurements.size()) and
+         "Source link index is outside the container bounds");
+
+  if (std::find(m_volumeIds.begin(), m_volumeIds.end(),
+                sourceLink.geometryId().volume()) == m_volumeIds.end()) {
+    m_fallback.calibrate(measurements, clusters, gctx, trackState);
+    return;
+  }
+
+  // TODO: Matrix size should be configurable perhaps?
+  std::array<float, 7 * 7> matrix =
+      ::detail::chargeMatrix((*clusters)[sourceLink.index()]);
+  std::vector<float> input;
+  input.assign(matrix.begin(), matrix.end());
+  input.push_back(sourceLink.geometryId().volume());
+  input.push_back(sourceLink.geometryId().layer());
+  input.push_back(trackState.parameters()[Acts::eBoundPhi]);
+  input.push_back(trackState.parameters()[Acts::eBoundTheta]);
+
+  std::visit(
+      [&](const auto& meas) {
+        auto E = meas.expander();
+        auto P = meas.projector();
+        Acts::ActsVector<Acts::eBoundSize> fpar = E * meas.parameters();
+        Acts::ActsSymMatrix<Acts::eBoundSize> fcov =
+            E * meas.covariance() * E.transpose();
+
+        input.push_back(fpar[Acts::eBoundLoc0]);
+        input.push_back(fpar[Acts::eBoundLoc1]);
+        input.push_back(fcov(Acts::eBoundLoc0, Acts::eBoundLoc0));
+        input.push_back(fcov(Acts::eBoundLoc1, Acts::eBoundLoc1));
+        if (input.size() != n_inputs) {
+          throw std::runtime_error("Expected input size of " +
+                                   std::to_string(n_inputs) +
+                                   ", got: " + std::to_string(input.size()));
+        }
+
+        std::vector<float> output = m_model.runONNXInference(input);
+        // Assuming 2-D measurements, the expected params structure is:
+        // [      0,    nComp[ --> priors
+        // [  nComp,  3*nComp[ --> means
+        // [3*nComp,  5*nComp[ --> variances
```

Comment:
Are the bracket backward or am I missing something ?
```suggestion
        // [      0,    nComp] --> priors
        // [  nComp,  3*nComp] --> means
        // [3*nComp,  5*nComp] --> variances
```

Reply:
They aren't, it's because the leftmost value is excluded. E.g. if nComp == 3, then the indices for the priors are 0, 1, 2 aka [0, 3[